geopandas/geodataframe.py

```diff
@@ -1391,8 +1411,8 @@ def __setitem__(self, key, value):
             if pd.api.types.is_scalar(value) or isinstance(value, BaseGeometry):
                 value = [value] * self.shape[0]
             try:
-                value = _ensure_geometry(value, crs=self.crs)
-                self._crs = value.crs
+                # TODO: use `crs=getattr(self, "crs", None)` in 0.12
+                value = _ensure_geometry(value, crs=self._crs)
```

Comment:
The following corner case would end up here:

```
df = geopandas.read_file(geopandas.datasets.get_path("naturalearth_lowres"))
geoms = list(df["geometry"])  # list of geoms without crs
# create a dataframe without geometry column, but currently still has cached _crs
df["geometry"] = 1
# assigning a list of geometry object will currently use _crs
df["geometry"] = geoms
```

So in the future, if we remove `_crs`, we will no longer preserve the crs here. So that's also a case to warn for? So maybe we can actually already use `crs=getattr(self, "crs", None)` as put in the comment, which will raise the warning (or trigger a custom warning here).

Reply:
This is probably such a corner case that it wasn't worth the effort to raise a warning for this, but I did so now anyway ..  (see last commit)